sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import javax.annotation.Nullable
+
+import scala.annotation.tailrec
+
+import org.apache.spark.sql.catalyst.analysis.TypeCoercion.hasStringType
+import org.apache.spark.sql.catalyst.expressions.{BinaryExpression, Cast, ComplexTypeMergingExpression, CreateArray, ExpectsInputTypes, Expression, Predicate, SortOrder}
+import org.apache.spark.sql.errors.QueryCompilationErrors
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{AbstractDataType, ArrayType, DataType, StringType}
+
+object CollationTypeCasts extends TypeCoercionRule {
+  override val transform: PartialFunction[Expression, Expression] = {
+    case e if !e.childrenResolved => e
+    case sc @ (_: BinaryExpression
+              | _: Predicate
+              | _: SortOrder
+              | _: ExpectsInputTypes
+              | _: ComplexTypeMergingExpression) =>
+      val newChildren = collateToSingleType(sc.children)
+      sc.withNewChildren(newChildren)
+    case pesc @ (_: CreateArray) =>
```

Comment:
why only `CreateArray` can preserve explicit collation?

Reply:
This might be a good question for @srielau. How do we want this query to behave: `SELECT array('a', 'b' collate UNICODE) || ('c' collate UNICODE_CI)` (Concat_Ws in our implementation)? This query in Postgre fails with explicit type mismatch. Do we want to propagate explicit collations from elements to arrays. I would say yes, as ArrayType is constructed with a parameter called `elementType`. This suggests, if elements are explicit, then ArrayType should be as well. Also, on other operations with arrays, as we defined that function outputs are implicit, I would say array creation is then the only place where we should propagate this explicit meaning.

Reply:
I'm not convinced. The propagation of "explicit collation" should be well-defined. Why don't funcitons like `substring`, `concat` propagate "explicit collation"?

Reply:
How I understood it, arrays are bulding blocks for types as well. The difference is that arrays have to be built from some different types which makes them complex. But `substring`, `concat`, etc. are functions that return something based on a transformation behaviour. In other words, CreateArray is not even considering the the values, it only builds a new 'basic' block. That is why I am also concerned with what we want to do with Elt, do we want to do casting or just propagation of inputs, as it does not change any string, but only takes a specific value from the given index. But again, this differs from array access, as arrays need to have a single type for all input expressions.

Reply:
Explicit collation NEVER propagates. It only applies to the immediate consumer.

Reply:
This is a good news. Let's not special-case CreateArray, otherwise people may question about CreateMap and CreateStruct. They are just SQL functions, not a special building block.

Reply:
We should play out how to the behavior.
If I want combine arrays or maps, how do I do that. Is this what drives us towards CAST support. And does that imply that teh result of a CAST with a COLLATE clause (on an element/key/value) is a explicit collation)?

Reply:
The example query is `SELECT array('a', 'b' collate UNICODE) || ('c' collate UNICODE_CI)`. This is kind of artificial as users can write one `array(...)` function, but think about a rewrite like `SELECT arr || ('c' collate UNICODE_CI) FROM (SELECT array('a', 'b' collate UNICODE) as arr)`, I think returning an array with string of collation UNICODE_CI is fine?

Reply:
Interesting. Yes array() sports an implicit collation. 'c' COLLATE UNICODE_CI sports an explicit collation.
The rules are clear that the explicit collation wins.
So the result is array<STRING UNICODE_CI> with an IMPLICIT collation.